include/thread_pool/safe_queue.h

```diff
@@ -382,8 +395,8 @@ class SafeQueue
     {
         if (this != &sq)
         {
-            std::lock_guard<std::mutex> lock1 (m_mutex);
-            std::lock_guard<std::mutex> lock2 (sq.m_mutex);
+            std::unique_lock<std::mutex> lock1 (m_mutex);
+            std::unique_lock<std::mutex> lock2 (sq.m_mutex);
```

Comment:
The change from lock_guard to unique_lock is probably only necessary here (with defer_lock), so that we can use the std::lock(lock1, lock2) to lock both the mutexes without deadlock. In all other functions with only a single lock, I think we can stick to keeping the lock_guard.